portfolio/src/main/java/com/google/sps/data/ArrayListTimeRangeGroup.java

```diff
@@ -0,0 +1,171 @@
+  }
+
+  /**
+   * Adds a new time range to the list. If the time range to add overlaps with any existing time
+   * range, the overlapping time ranges will be merged.
+   */
+  public void addTimeRange(TimeRange timeRange) {
+    // If the original allTimeRanges list is empty,
+    // this is the first time we add anything to the list,
+    // so simply add the new time range to the list and return.
+    if (allTimeRanges.isEmpty()) {
+      allTimeRanges.add(timeRange);
+      return;
+    }
+
+    List<TimeRange> newTimeRanges = new ArrayList<TimeRange>();
+
+    // This variable represents the latest time range previously exmained.
+    // Initially, this variable points to the time range we want to add.
+    // As the for loop iterates through all the current time ranges already existing
+    // in allTimeRanges, this lastExaminedTimeRange variable is the time range with which
+    // any existing time range merges with, if necessary.
+    TimeRange lastExaminedTimeRange = timeRange;
+
+    for (int i = 0; i < allTimeRanges.size(); i++) {
+      TimeRange currentRange = allTimeRanges.get(i);
+
+      // If the current range is complete contained by the lastExaminedTimeRange,
+      // no merging or adding needs to happen.
+      if (lastExaminedTimeRange.contains(currentRange)) {
+        continue;
+      }
+
+      // The case for when two time ranges need to be merged.
+      // Change the lastExaminedTimeRange pointer to the new time range created
+      // after the merging.
+      if (lastExaminedTimeRange.overlaps(currentRange)) {
+        lastExaminedTimeRange = mergeTwoTimeRanges(currentRange, lastExaminedTimeRange);
+      } else {
+        // This is the case that lastExaminedTimeRange and the current time range do not overlap.
+        // If the current range from the original list ends after the last examined time range,
+        // add the time range pointed to by lastExaminedTimeRange to the new list of all ranges, and
+        // change the lastExaminedTimeRange to point to the current range.
+        if (currentRange.end().isAfter(lastExaminedTimeRange.end())) {
+          newTimeRanges.add(lastExaminedTimeRange);
+          lastExaminedTimeRange = currentRange;
+        } else {
+          // If lastExaminedTimeRange is later than the current range,
+          // add the current range to the new list of all ranges.
+          // The lastExaminedTimeRange pointer shouldn't move.
+          newTimeRanges.add(currentRange);
+        }
+      }
+
+      // If current time range is the last element in the allTimeRanges list,
+      // add the last examined time range to the new list.
+      if (i == allTimeRanges.size() - 1) {
+        newTimeRanges.add(lastExaminedTimeRange);
+      }
+    }
+
+    // Finally, set the global variable allTimeRanges to this newly built list of time ranges.
+    allTimeRanges = newTimeRanges;
+    Collections.sort(allTimeRanges, TimeRange.SORT_BY_TIME_RANGE_START_TIME_ASCENDING);
```

Comment:
I don't think this call is necessary, if we ensure that when this method is invoked the internal list is already sorted.